### PR TITLE
Added events to SFX command

### PR DESCRIPTION
**Pull request by jherbage-splunk**
This PR adds the ability to forward SPL events as SignalFX events (as opposed to the existing metrics tosfx command).
Also added unit tests for it and a fake handler for /events EP in SFX.
Cleaned up some of the tests file to make the unit testing to work and added instructions for unit testing (thanks Jeffrey Cheng) to the README
Note that eventstosfx has been skipped from pylint in circleci tests because it does dup0licate some tosfx code.
I did try to refactor to a base class for both but the its limiting because we are already using the Splunk base classes so not sure it can be done.
circleci tests all passed on this fork and pytest unit testing also passed

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,6 +10,34 @@ See [./signalfx-forwarder-app/README.md](./signalfx-forwarder-app/README.md) for
...
 Run `slim package signalfx-forwarder` to generate the .tar.gz and upload to the Github release.
 
+#### CI testing
+
+The repo includes a set of circleci tests that need to pass before your Pull Request can be merged.
+Perform the tests here https://app.circleci.com/ pointed at your updated github fork.
+
+#### Unit Testing
+Unit testing should be completed successfully and can be performed locally as follows.
+If you are adding new commands you may need to updated the test cases defined in [./signalfx-forwarder-app/tests/splunk_test.py](./signalfx-forwarder-app/tests/splunk_test.py) and possibly the fake handling of the commands [./signalfx-forwarder-app/tests/helpers/fake_backend.py](./signalfx-forwarder-app/tests/helpers/fake_backend.py)
```
Comment (rmfitzpatrick) [inline README.md:20]:
```suggestion
If you are adding new commands you may need to update the test cases defined in [./signalfx-forwarder-app/tests/splunk_test.py](./signalfx-forwarder-app/tests/splunk_test.py) and possibly the fake handling of the commands [./signalfx-forwarder-app/tests/helpers/fake_backend.py](./signalfx-forwarder-app/tests/helpers/fake_backend.py)
```

File: signalfx-forwarder-app/README.md [hunk 2]
```diff
@@ -22,6 +22,9 @@ A custom search command to metricize Splunk events and send them to SignalFx
 **tosfx**:       Send the datapoints to SignalFx in a non streaming manner.
                  Does not work in realtime mode, but has higher throughput.
 
+**eventstosfx**:       Send the datapoints to SignalFx as events.
```
Comment (rmfitzpatrick) [inline signalfx-forwarder-app/README.md:25]: `tosfxevents` since they are the final format?

File: signalfx-forwarder-app/app.manifest [hunk 3]
```diff
@@ -5,7 +5,7 @@
     "id": {
       "group": null, 
       "name": "signalfx-forwarder-app", 
-      "version": "2.0.1"
+      "version": "2.0.2"
```
Comment (rmfitzpatrick) [inline signalfx-forwarder-app/app.manifest:8]: Not totally sure of the versioning practices here, but a new feature would warrant `2.1.0` imo.

File: signalfx-forwarder-app/bin/eventstosfx.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
+# pylint: skip-file
```
Comment (rmfitzpatrick) [inline signalfx-forwarder-app/bin/eventstosfx.py:1]: Sure this is desired?

File: signalfx-forwarder-app/bin/eventstosfx.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+    SPLUNK_PASSWORD_REALM = "realm"
+    SPLUNK_PASSWORD_USER_NAME = "username"
+    SPLUNK_PASSWORD_CLEAR_PASSWORD = "clear_password"
+    SPLUNK_KV_STORE_SFX_CONFIG_COLLECTION_NAME = "sfx_ingest_config"
+    SPLUNK_SFX_CONFIG_INGEST_URL_KEY = "ingest_url"
+    SPLUNK_PASSWORDS_STORAGE_SFX_ACCESS_TOKEN_REALM = "sfx_ingest_command"
+    SPLUNK_PASSWORDS_STORAGE_SFX_ACCESS_TOKEN_USER_NAME = "access_token"
+
+    def ensure_default_config(self):
```
Comment (rmfitzpatrick) [inline signalfx-forwarder-app/bin/eventstosfx.py:53]: nit, this method doesn't appear to do what the name suggests.  Maybe something like `ensure_connection_info()`?

File: signalfx-forwarder-app/bin/eventstosfx.py (hunk 4 above)
Comment (jherbage-splunk) [reply signalfx-forwarder-app/bin/eventstosfx.py:1]: Yeah defintiely agree we want it but the lint is complaining that the code in tosfx and eventstosfx is duplicated and shoulld be refactored. I did tryi refactoring to a super class that inherits EventingCommand but it wasnt working and I ran out of time - hence just ignoring the lint.

File: signalfx-forwarder-app/bin/eventstosfx.py (hunk 5 above)
Comment (jherbage-splunk) [reply signalfx-forwarder-app/bin/eventstosfx.py:53]: I copied the original naming convention from whoever put tosfx.py in

File: signalfx-forwarder-app/app.manifest (hunk 3 above)
Comment (jherbage-splunk) [reply signalfx-forwarder-app/app.manifest:8]: Agreed - will push with a new version

File: signalfx-forwarder-app/README.md (hunk 2 above)
Comment (jherbage-splunk) [reply signalfx-forwarder-app/README.md]: good point and changed now

File: signalfx-forwarder-app/bin/tosfxevents.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            )
+            for event in out:
+                event["status"] = resp.status_code
+                if resp.status_code != 200:
+                    event["response_error"] = resp.content
+
+        for event in out:
+            yield event
+
+def send_payload(payload, target_url, token):
```
Comment (rmfitzpatrick) [inline signalfx-forwarder-app/bin/tosfxevents.py:116]: Can the existing function in tosfx not be used?

File: signalfx-forwarder-app/bin/tosfxevents.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        data=body.read(),
+    )
+    return resp
+    
+def compose_ingest_url(ingest_base_url, ev_endpoint):
+    return ingest_base_url.rstrip("/") + ev_endpoint
+
+def add_event_to_payload(self, event, payload):
+
+    eventDict = dict()
```
Comment (rmfitzpatrick) [inline signalfx-forwarder-app/bin/tosfxevents.py:134]: overwritten below and not used?

File: signalfx-forwarder-app/bin/tosfxevents.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    dimensions = dict()
+    properties = dict()
+    timestamp = None
+    eventType = None
+
+
+    for key, value in event.items():
+        self.logger.error("key is "+str(key) + " value is "+str(value))
```
Comment (rmfitzpatrick) [inline signalfx-forwarder-app/bin/tosfxevents.py:142]: error logging for all items seems undesirable.  Is debug logging an option?

File: signalfx-forwarder-app/bin/tosfxevents.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    ## Syntax
+
+    <command> | tosfxevents
+
+    ## Description
+
+    One or more datapoints are generated for each input event's field(s) of the
+    form `gauge_*`, `counter_*` or `cumulative_counter_*`.  The metric name in
+    SignalFx will be the `*` part of the field name.  Any additional fields on
+    the event will be attached as dimensions to the generated datapoints.
```
Comment (rmfitzpatrick) [inline signalfx-forwarder-app/bin/tosfxevents.py:35]: Do you mind updating this description with the added functionality?

File: README.md [hunk 10]
```diff
@@ -10,6 +10,34 @@ See [./signalfx-forwarder-app/README.md](./signalfx-forwarder-app/README.md) for
 You'll need to install the [Splunk Packacking Toolkit](https://dev.splunk.com/enterprise/docs/releaseapps/packagingtoolkit/installpkgtoolkit).
 Run `slim package signalfx-forwarder` to generate the .tar.gz and upload to the Github release.
 
+#### CI testing
+
+The repo includes a set of circleci tests that need to pass before your Pull Request can be merged.
+Perform the tests here https://app.circleci.com/ pointed at your updated github fork.
+
+#### Unit Testing
```
Comment (rmfitzpatrick) [inline README.md:18]: Thanks for adding all of this.

**Comment by rmfitzpatrick** [comment]
@jherbage-splunk I think there's an issue with Circle CI when it's configured to follow forked projects.  In the past I've had to "stop building" my forks in order for the checks to run against upstream PRs (https://app.circleci.com/pipelines/github/jherbage-splunk/splunk-forwarder)

File: signalfx-forwarder-app/bin/tosfxevents.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+    dimensions = dict()
+    properties = dict()
+    timestamp = None
+    event_type = None
+
+
+    for key, value in event.items():
+        if value != "":
+            if key.startswith("event_"):
+                event_type = value
```
Comment (rmfitzpatrick) [inline signalfx-forwarder-app/bin/tosfxevents.py:144]: wonder if dropping the `event_` would be desired

File: signalfx-forwarder-app/bin/tosfxevents.py [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+            elif key.startswith("property_"):
+                if value[0] != "_" and len(value) < 256:
+                    new_key = key.replace("property_", "")
+                    new_key = new_key.replace(".", "_")
```
[the remark this answers is not in the document]
Comment (mstumpfx) [reply signalfx-forwarder-app/bin/tosfxevents.py:148]: Eh apparently this works in weird nuanced ways after testing more. Disregard.

**Review by mstumpfx — APPROVED** [review]
@keitwb did you want to look this look this over once?

File: signalfx-forwarder-app/README.md [hunk 13]
```diff
@@ -44,6 +47,10 @@ These arguments can be appended to the new commands. For example, `tosfx debug=t
 index=_internal group=per_index_thruput series!=_* | rename ev AS eventCount | rename kb AS kilobytes | table _time kilobytes eventCount series host | `gauge(kilobytes)` |`gauge(eventCount)` | tosfx
 ```
 
+```
+index=_internal group=per_index_thruput series!=_* | rename ev AS eventCount | rename kb AS kilobytes | table _time kilobytes eventCount series host | `gauge(kilobytes)` |`gauge(eventCount)` | tosfxevents
```
Comment (keitwb) [inline signalfx-forwarder-app/README.md:51]: This example is not very illustrative of a practical use case for this.

**Pull request merged**